%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: sync
components:
- name: left_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_image_receiver
    min_size: 1
- name: right_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_image_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- type: nvidia::gxf::Synchronization
  parameters:
    inputs: [left_image_receiver, right_image_receiver, left_cam_receiver, right_cam_receiver]
    outputs: [left_image_transmitter, right_image_transmitter, left_cam_transmitter, right_cam_transmitter]
- name: left_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_image_transmitter
    min_size: 1
- name: right_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_image_transmitter
    min_size: 1
- name: left_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_cam_transmitter
    min_size: 1
- name: right_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_cam_transmitter
    min_size: 1
---
name: ess
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- name: confidence_output
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: confidence_output
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 2211840
    num_blocks: 4
- name: cuda_stream
  type: nvidia::gxf::CudaStreamPool
  parameters:
    stream_flags: 1
    stream_priority: 0
    reserved_size: 1
    max_size: 1
- type: nvidia::isaac::ESSInference
  parameters:
    output_name: frame
    pool: pool
    left_image_receiver: data_receiver_left
    right_image_receiver: data_receiver_right
    output_transmitter: data_transmitter
    confidence_transmitter: confidence_output
    image_type: RGB_U8
    pixel_mean: [-128, -128, -128]
    normalization: [0.00392156862, 0.00392156862, 0.00392156862]
    standard_deviation: [0.5, 0.5, 0.5]
    max_batch_size: 1
    model_input_type: RGB_U8
    onnx_file_path: onnx_file_path_placeholder
    engine_file_path: engine_file_path_placeholder
    input_layers_name: [input_left, input_right]
    output_layers_name: [output_left, output_conf]
    preprocess_type: RESIZE
    stream_pool: cuda_stream
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: pool
    min_blocks: 2
---
name: disparity_thresholder
components:
- name: image_input
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: image_input
    min_size: 1
- name: confidence_input
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: confidence_input
    min_size: 1
- name: masked_output
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: masked_output
    min_size: 1
- name: allocator
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 2211840
    num_blocks: 2
- name: cuda_stream
  type: nvidia::gxf::CudaStreamPool
  parameters:
    stream_flags: 1
    stream_priority: 0
    reserved_size: 1
    max_size: 1
- name: image_thresholder
  type: nvidia::isaac::VideoBufferThresholder
  parameters:
    image_input: image_input
    video_buffer_name: frame
    mask_input: confidence_input
    masked_output: masked_output
    allocator: allocator
    fill_value_float: -1.0
    stream_pool: cuda_stream
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: allocator
    min_blocks: 1
---
name: disparity_compositor
components:
- name: disparity_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: disparity_in
    min_size: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- type: nvidia::isaac::DisparityCompositor
  parameters:
    left_camera_model_receiver: left_cam_receiver
    right_camera_model_receiver: right_cam_receiver
    disparity_receiver: disparity_in
    output_transmitter: data_transmitter
    min_disparity: 0
    max_disparity: -1
---
name: image_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: camera_info_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: throttler
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- name: data_transmitter_left
  type: nvidia::gxf::DoubleBufferTransmitter
- name: data_transmitter_right
  type: nvidia::gxf::DoubleBufferTransmitter
- name: left_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
- name: right_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter_left
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter_right
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_cam_transmitter
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_cam_transmitter
    min_size: 1
- name: throttler1
  type: nvidia::isaac::MultiChannelsThrottler
  parameters:
    inputs: [data_receiver_left, data_receiver_right]
    outputs: [data_transmitter_left, data_transmitter_right]
- name: throttler2
  type: nvidia::isaac::MultiChannelsThrottler
  parameters:
    inputs: [left_cam_receiver, right_cam_receiver]
    outputs: [left_cam_transmitter, right_cam_transmitter]
---
name: splitter
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- name: image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: image_transmitter
    min_size: 1
- name: camera_model_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_model_transmitter
    min_size: 1
- name: splitter
  type: nvidia::isaac_ros::MessageSplitter
  parameters:
    rx: data_receiver
    image_tx: image_transmitter
    camera_model_tx: camera_model_transmitter
---
name: connections
components:
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_image_transmitter
    target: throttler/data_receiver_left
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_image_transmitter
    target: throttler/data_receiver_right
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_cam_transmitter
    target: throttler/left_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_cam_transmitter
    target: throttler/right_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: throttler/data_transmitter_left
    target: ess/data_receiver_left
- type: nvidia::gxf::Connection
  parameters:
    source: throttler/data_transmitter_right
    target: ess/data_receiver_right
- type: nvidia::gxf::Connection
  parameters:
    source: throttler/left_cam_transmitter
    target: disparity_compositor/left_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: throttler/right_cam_transmitter
    target: disparity_compositor/right_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: ess/data_transmitter
    target: disparity_thresholder/image_input
- type: nvidia::gxf::Connection
  parameters:
    source: ess/confidence_output
    target: disparity_thresholder/confidence_input
- type: nvidia::gxf::Connection
  parameters:
    source: disparity_thresholder/masked_output
    target: disparity_compositor/disparity_in
- type: nvidia::gxf::Connection
  parameters:
    source: disparity_compositor/data_transmitter
    target: splitter/data_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: splitter/image_transmitter
    target: image_sink/signal
- type: nvidia::gxf::Connection
  parameters:
    source: splitter/camera_model_transmitter
    target: camera_info_sink/signal
---
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::EventBasedScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
    worker_thread_number: 4
- type: nvidia::gxf::JobStatistics
  parameters:
    clock: clock
